### PR TITLE
Make game.size take zoom into consideration

### PR DESCRIPTION
**Pull request by spydon**
# Description

Since `game.size` wasn't taking zoom into consideration `followComponent` didn't work when zoom was anything else than 1.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] My PR does not decrease the code coverage, or I have __a very special case__ and explained on the PR description why this PR decreases the coverage.
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `./scripts/format.sh` and the Flame analyzer (`./scripts/analyze.sh`) does not report any problems.
- [x] I read and followed the [Flame Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [x] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [ ] No, this is *not* a breaking change.

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

File: doc/camera_and_viewport.md [hunk 1]
```diff
@@ -29,17 +29,17 @@ your needs):
 
 When using `BaseGame`, the operations performed by the viewport are done automatically to every
 render operation, and the `size` property in the game, instead of the logical widget size, becomes
-the size as seen through the viewport. If for some reason you need to access the original real
-logical pixel size, you can use `canvasSize`. For a more in depth description on what each Viewport
-does and how it operates, check the documentation on its class.
+the size as seen through the viewport together with the zoom of the camera. If for some reason you
+need to access the original real logical pixel size, you can use `canvasSize`. For a more in depth
+description on what each Viewport does and how it operates, check the documentation on its class.
```
Comment (wolfenrain) [inline doc/camera_and_viewport.md:34]:
```suggestion
description on what each `Viewport` does and how it operates, check the documentation on its class.
```

File: doc/camera_and_viewport.md [hunk 2]
```diff
@@ -29,17 +29,17 @@ your needs):
...
 
 ## Camera
 
 Unlike the Viewport, the Camera is a more dynamic Canvas transformation that is normally dependent
```
Comment (wolfenrain) [inline doc/camera_and_viewport.md:38]:
```suggestion
Unlike the `Viewport`, the `Camera` is a more dynamic `Canvas` transformation that is normally dependent
```

Not part of your PR, but I feel this sentence is not correctly worded? "is a more ... that is normally dependent on" and then a list?

File: doc/camera_and_viewport.md [hunk 3]
```diff
@@ -29,17 +29,17 @@ your needs):
...
 on:
 
  * world coordinates that do not match screen coordinates 1:1
```
Comment (wolfenrain) [inline doc/camera_and_viewport.md:41]: Should these bulletins not end with either a `.` or `;`?

File: packages/flame/CHANGELOG.md [hunk 4]
```diff
@@ -18,7 +18,7 @@
  - Refactor TextBoxComponent
  - Fix bugs with TextBoxComponent
  - Improve error message for composed components
-
+ - Fix game.size to take zoom into consideration (which also fixes followComponent when zoom != 1)
```
Comment (wolfenrain) [inline packages/flame/CHANGELOG.md:21]:
```suggestion
 - Fix `game.size` to take zoom into consideration (which also fixes `camera.followComponent` when `zoom != 1`)
```

File: packages/flame/lib/src/game/base_game.dart [hunk 5]
```diff
@@ -68,12 +68,15 @@ class BaseGame extends Game with FPSCounter {
   /// This is overwritten to consider the viewport transformation.
   ///
   /// Which means that this is the logical size of the game screen area as
-  /// exposed to the canvas after viewport transformations.
+  /// exposed to the canvas after viewport transformations and camera zooming.
```
Comment (wolfenrain) [inline packages/flame/lib/src/game/base_game.dart:71]: These dartdocs are not available on `size` if you put them on the `_sizeBuffer`. And as this dartdoc starts with `This is overwritten..` I think we should move this whole block down to `size`...

File: packages/flame/lib/src/game/camera.dart [hunk 6]
```diff
@@ -203,6 +203,18 @@ class Camera extends Projector {
     return worldCoordinates * zoom;
   }
 
+  /// Takes coordinates in the screen space and returns their counter-part in
+  /// the world space
```
Comment (wolfenrain) [inline packages/flame/lib/src/game/camera.dart:207]:
```suggestion
  /// the world space.
```

File: packages/flame/lib/src/game/camera.dart [hunk 7]
```diff
@@ -203,6 +203,18 @@ class Camera extends Projector {
...
+  Vector2 screenToWorld(Vector2 screenCoordinates) {
+    return unprojectVector(screenCoordinates);
+  }
+
+  /// Takes coordinates in the world space and returns their counter-part in
+  /// the screen space
```
Comment (wolfenrain) [inline packages/flame/lib/src/game/camera.dart:213]:
```suggestion
  /// the screen space.
```

File: packages/flame/lib/src/game/camera.dart [hunk 8]
```diff
@@ -203,6 +203,18 @@ class Camera extends Projector {
...
+  Vector2 worldToScreen(Vector2 worldCoordinates) {
```
Comment (wolfenrain) [inline packages/flame/lib/src/game/camera.dart:214]: These could be a one-liner? Not sure about the rules for it.. Just a suggestion :D

**Review by wolfenrain — APPROVED** [review]

File: packages/flame/lib/src/game/camera.dart (hunk 8 above)
Comment (spydon) [reply packages/flame/lib/src/game/camera.dart:214]: Nicer to have all these conform to the same style I think

File: packages/flame/lib/src/game/base_game.dart (hunk 5 above)
Comment (spydon) [reply packages/flame/lib/src/game/base_game.dart:72]: Fixed

File: packages/flame/lib/src/game/camera.dart (hunk 8 above)
Comment (wolfenrain) [reply packages/flame/lib/src/game/camera.dart:214]: We should document it in our styleguide then :D

File: packages/flame/lib/src/game/base_game.dart [hunk 9]
```diff
@@ -65,15 +65,18 @@ class BaseGame extends Game with FPSCounter {
 
   late Projector _combinedProjector;
 
+  final Vector2 _sizeBuffer = Vector2.zero();
   /// This is overwritten to consider the viewport transformation.
   ///
   /// Which means that this is the logical size of the game screen area as
-  /// exposed to the canvas after viewport transformations.
+  /// exposed to the canvas after viewport transformations and camera zooming.
```
Comment (wolfenrain) [inline packages/flame/lib/src/game/base_game.dart:72]:
```suggestion
  /// exposed to the canvas after viewport transformations and camera zooming.
  ///
```

**Review by wolfenrain — APPROVED** [review]

File: packages/flame/lib/src/game/camera.dart (hunk 8 above)
Comment (spydon) [reply packages/flame/lib/src/game/camera.dart:214]: I just mean the same style as for the `project/unproject` methods, I think this is a bit too specific to have in the styleguide. :)

Comment (wolfenrain) [reply packages/flame/lib/src/game/camera.dart:214]: Oh I was talking about the expression methods vs block methods :D

File: packages/flame/CHANGELOG.md [hunk 10]
```diff
@@ -18,9 +18,9 @@
  - Refactor TextBoxComponent
  - Fix bugs with TextBoxComponent
  - Improve error message for composed components
+ - Fix `game.size` to take zoom into consideration (which also fixes `camera.followComponent` when `zoom != 1`)
```
Comment (erickzanardo) [inline packages/flame/CHANGELOG.md:21]: I would split this into two entries, one for the fix on game.size and other for the followcomponent fix

**Review by erickzanardo — APPROVED** [review]

File: packages/flame/CHANGELOG.md (hunk 10 above)
Comment (spydon) [reply packages/flame/CHANGELOG.md]: Done!

**Review by wolfenrain — APPROVED** [review]
LGTM

**Review by wolfenrain — APPROVED** [review]

File: packages/flame/lib/src/components/mixins/collidable.dart [hunk 11]
```diff
@@ -26,28 +26,18 @@ class ScreenCollidable extends PositionComponent
...
-
-  void _updateSize() {
-    if (_effectiveSize != gameRef.viewport.effectiveSize ||
-        _zoom != gameRef.camera.zoom) {
-      _effectiveSize.setFrom(gameRef.viewport.effectiveSize);
-      _zoom = gameRef.camera.zoom;
-      size = _effectiveSize / _zoom;
-    }
+    position = gameRef.camera.unprojectVector(_zeroVector);
+    size = gameRef.size;
```
Comment (luanpotter) [inline packages/flame/lib/src/components/mixins/collidable.dart:41]: I guess this is a very textbook only example, as I can't imagine a game that uses camera but things collide with the actual screen size, and not something fixed in the world right? so having to carry over like this should be rare.

**Review by luanpotter — APPROVED** [review]
LGTM

**Pull request merged**